lua/bufferline/jump_mode.lua

```diff
@@ -20,36 +20,41 @@ local state = require'bufferline.state'
 -- Section: Buffer-picking mode state --
 ----------------------------------------
 
-local M = {
+--- @class bufferline.JumpMode
+--- @field private buffer_letter_bimap table<integer|string, integer|string> a bi-directional map of buffer integers and their letters.
+--- @field private index_by_letter table<string, integer> `letters` in the order they were provided
+--- @field private letter_status table<integer, boolean>
+--- @field private letters string the letters to allow while jumping
+--- @field private reinitialize boolean whether to run `initialize_indexes` on next jump.
+local JumpMode = {
   letters = vim.g.bufferline.letters, -- array
   index_by_letter = {}, -- object
   letter_status = {}, -- array
-  buffer_by_letter = {}, -- object
-  letter_by_buffer = {}, -- object
+  buffer_letter_bimap = {}, -- object
+  reinitialize = false, -- bool
```

Comment:
I opted to lazily reassign the buffer letters, because a user may free more than one buffer letter before entering jump mode again (if they enter jump mode again at all). This avoids the performance decrease which would be associated with shuffling the indices every time a buffer is closed.